### ISSUE TITLE
Add `docker pull` retries

### ISSUE THREAD

**Issue by phlax** [envoyproxy/envoy#25545]
We currently get CI flakes due to connectivity issues with dockerhub (mostly resutling in an `EOF` error)

We should try to limit how much we need to pull - but we can also mitigate with some retries

We dont want to flood dockerhub with requests if its already failing so we want to limit this - also if dockerhub is failing and we add too many retries with long intervals it would slow already failing ci down

That said i think we can make CI more robust - especially in the examples, but probably pulling the build image also

I reckon a retry count/interval of 3/10s would be reasonable

**Comment by phlax** [comment]
in the case of the examples we could possibly make retrying conditional on no previous tests failing

**Comment by daixiang0** [comment]
hh, I remember we discuss this before.

How about download the base image as cache?

**Comment by phlax** [comment]
> hh, I remember we discuss this before.

yep, for sure - i think you may have raised a PR - i was gonna check for it

> How about download the base image as cache?

that would be the idea, but contrary to what the azure docs say it doesnt make it faster for the (now ~7GB) build image - it makes it much slower

i did manage to get a speed improvement doing this - but only by using ramdisks to download the image before loading it into docker

the problem with anything involving azp cache is that it is so unreliable - above comments are for normal circumstances, but even with current setup some of our caches can take 20 mins on one run and then back to ~normal the next

**Comment by phlax** [comment]
experimenting with this again, i can get this down to ~1m5s (so far) by copying the cache to a ramdisk and installing (by physically replacing `/var/lib/docker` from there).

im not sure if this is faster than dockerhub - i suspect a bit slower

the problem with azp cache unreliability seems to be known and can resolved by adding an env var to workaround (worked so far)

https://github.com/microsoft/azure-pipelines-agent/pull/4077

**Comment by phlax** [comment]
this should be resolved by #26053 

we can add retries on examples if necessary but im more looking at caching the `/var/lib/docker` for that first
